sycl/source/detail/composite_device/composite_device.cpp

```diff
@@ -0,0 +1,42 @@
+//==---------- composite_device.cpp - SYCL Composite Device ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <sycl/ext/oneapi/experimental/composite_device.hpp>
+#include <sycl/platform.hpp>
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::oneapi::experimental {
+std::vector<device> get_composite_devices() {
+  std::vector<device> Composites;
+  auto Devs = sycl::device::get_devices();
+  for (const auto &D : Devs) {
+    if (D.has(sycl::aspect::ext_oneapi_is_component)) {
+      auto Composite = D.get_info<info::device::composite_device>();
+      // Filter out duplicates.
+      if (std::find(Composites.begin(), Composites.end(), Composite) ==
+          Composites.end())
+        Composites.push_back(Composite);
+    }
+  }
+  std::vector<device> Result;
+  for (const auto &Composite : Composites) {
+    auto Components = Composite.get_info<info::device::component_devices>();
+    // Only return composite devices if all of its component devices are
+    // available.
+    if (std::all_of(Components.begin(), Components.end(), [&](const device &d) {
+          return std::find(Devs.begin(), Devs.end(), d) != Devs.end();
+        })) {
+      Result.push_back(Composite);
+    }
+  }
+  return Result;
```

Comment:
That can be an `std::copy_if(Composites.begin(), Composites.end(), std::back_inserter{Result}, [](...) { /* predicate */ });`

Reply:
Updated in https://github.com/intel/llvm/pull/12178/commits/8d68661b1c4a6e35ea6ab80d0b5404071bbac958.